src/simulator.rs

```diff
@@ -44,7 +44,11 @@ impl Simulator {
         for src in 0..self.elements.len() {
             for (rpc, dst) in &self.rpc_buffer[src] {
                 if (*dst).is_some() {
-                    self.elements[(*dst).unwrap() as usize].recv(*rpc, tick);
+                    // Before we send this rpc on, we should update its path to include the most recently traversed element
```

Comment:
If you access the object "rpc" mutably, there should be a way to modify it in place. You can use the mut keyword for this: https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html

Reply:
Yes, I have tried to access rpc mutably, and that part works.  The issue is that rpc has an element, path, of type String, and String does not implement copy.  Path is of type String because it is updated at every hop, and so needs to be mutable.  The other possible string option, str, isn't mutable.

If I add to the path using a mutable version of rpc, then when I go to do
self.elements[(*dst).unwrap() as usize].recv(*rpc, tick);
I can't.  Dereferencing is the same thing as copying the object over, and copy isn't implemented for String, so it throws an error:

cannot move out of `*rpc` which is behind a mutable reference
move occurs because `*rpc` has type `Rpc`, which does not implement the `Copy` trait

I'm not sure if there's a clean way to make path easily editable, but also allow rpc to be dereferenced/copied easily.

Reply:
OK. For now, let's go with the clone() statement and keep the TODO.